p751toolbox/curve.go

```diff
@@ -109,6 +37,120 @@ type ProjectivePrimeFieldPoint struct {
+	t0.Sub(&t0, &t1)      // t0 = t0 - t1
+	j.Sub(&t0, &t1)       // t0 = t0 - t1
+	t1.Square(&t1)        // t1 = t1^2
+	j.Mul(&j, &t1)        // t0 = t0 * t1
+	t0.Add(&t0, &t0)      // t0 = t0 + t0
+	t0.Add(&t0, &t0)      // t0 = t0 + t0
+	t1.Square(&t0)        // t1 = t0^2
+	t0.Mul(&t0, &t1)      // t0 = t0 * t1
+	t0.Add(&t0, &t0)      // t0 = t0 + t0
+	t0.Add(&t0, &t0)      // t0 = t0 + t0
+	j.Inv(&j)             // j  = 1/j
+	j.Mul(&t0, &j)        // j  = t0 * j
+
+	j.ToBytes(jBytes)
+}
+
+// Given affine points x(P), x(Q) and x(Q-P) in a extension field F_{p^2}, function
+// recorvers projective coordinate A of a curve. This is Algorithm 10 from SIKE.
+func (curve *ProjectiveCurveParameters) RecoverCoordinateA(xp, xq, xr *ExtensionFieldElement) {
+	var t0, t1 ExtensionFieldElement
+
+	t1.Add(xp, xq)                            // t1 = Xp + Xq
+	t0.Mul(xp, xq)                            // t0 = Xp * Xq
+	curve.A.Mul(xr, &t1)                      // A  = X(q-p) * t1
+	curve.A.Add(&curve.A, &t0)                // A  = A + t0
+	t0.Mul(&t0, xr)                           // t0 = t0 * X(q-p)
+	curve.A.Sub(&curve.A, &oneExtensionField) // A  = A - 1
+	t0.Add(&t0, &t0)                          // t0 = t0 + t0
+	t1.Add(&t1, xr)                           // t1 = t1 + X(q-p)
+	t0.Add(&t0, &t0)                          // t0 = t0 + t0
+	curve.A.Square(&curve.A)                  // A  = A^2
+	t0.Inv(&t0)                               // t0 = 1/t0
+	curve.A.Mul(&curve.A, &t0)                // A  = A * t0
+	curve.A.Sub(&curve.A, &t1)                // A  = A - t1
+}
+
+// Computes equivalence (A:C) ~ (A+2C : A-2C)
+func (curve *ProjectiveCurveParameters) CalcCurveParamsEquiv3() CurveCoefficientsEquiv {
+	var coef CurveCoefficientsEquiv
+	var tmp ExtensionFieldElement
+
+	// TODO: Calling code sets C=1, always (all functions). Currently only tests
+	//       require C to be customizable.
+
+	// C24 = 2*C
+	tmp.Add(&curve.C, &curve.C)
+	// A24_plus = A + 2C
+	coef.A.Add(&curve.A, &tmp)
+	// A24_minus = A - 2C
+	coef.C.Sub(&curve.A, &tmp)
+	return coef
+}
+
+// Computes equivalence (A:C) ~ (A+2C : 2C)
+func (cparams *ProjectiveCurveParameters) CalcCurveParamsEquiv4() CurveCoefficientsEquiv {
+	var coefEq CurveCoefficientsEquiv
+	// C = 2*cparams.C
+	coefEq.C.Add(&cparams.C, &cparams.C)
+	// A24_plus = A + 2C
+	coefEq.A.Add(&cparams.A, &coefEq.C)
+	// C24 = 4*C
+	coefEq.C.Add(&coefEq.C, &coefEq.C)
+	return coefEq
+}
+
+// Helper function for RightToLeftLadder(). Returns A+2C / 4.
+func (cparams *ProjectiveCurveParameters) calcAplus2Over4() (ret ExtensionFieldElement) {
+	var tmp ExtensionFieldElement
+	// 2C
+	tmp.Add(&cparams.C, &cparams.C)
+	// A+2C
+	ret.Add(&cparams.A, &tmp)
+	// 1/4C
+	tmp.Add(&tmp, &tmp).Inv(&tmp)
+	// A+2C/4C
+	ret.Mul(&ret, &tmp)
+	return
+}
+
+// Recovers (A:C) curve parameters from projectively equivalent (A+2C:A-2C).
+func (cparams *ProjectiveCurveParameters) RecoverCurveCoefficients3(coefEq *CurveCoefficientsEquiv) {
```

Comment:
This seems to take (A+2C:A-2C) and return (4A : 3A +2C) instead of (A:C)

Contrast this with RecoverCurveCoefficients4() below which actually inverts CalcCurveParamsEquiv4()

Reply:
nop, it does:

cparams.A = 2*(coefEq.A + coefEq.C)
cparams.C = (coefEq.A + coefEq.C)

Substituting coefEq.A with A+2C and coefEq.C with A-2C gives:
cparams.A = 4A
cparams.C = 4C

Hence,
cparams.A/cparams.C = A/C